### ISSUE TITLE
belongsToCommunity has an GenericFile Id instead of Collection Id

### ISSUE THREAD

**Issue by weiweishi** [ualbertalib/HydraNorth#712]
With the issue identified by @sfarnel on the comment of #710:
e.g.: https://hydranorth.library.ualberta.ca/files/dn39x224x

visibility shows as UofA. when I click on an item I get an error:

"We're sorry, but something went wrong. If you are the application owner
check the logs for more information."

Log info:
ActionView::Template::Error (Model mismatch. Expected Collection. Got: GenericFile):
     8:         <% @generic_file.belongsToCommunity.each do |c| %>
     9:         <tr class="collection attributes">
    10:           <%= content_tag :td, class: 'attribute title', data: { id: c } do %>
    11:             <% community = Collection.find(c) %>
    12:             <%= link_to community.title, collections.collection_path(community) %>
    13:           <% end %>
    14:         </tr>

Some issue with a particular item:
uuid:ca304e6a-06ab-419d-aa52-1771d3b641ee

The community in the foxml has an uuid:11274e20-0426-4e80-84f4-bef79dbd6633 
However, this is not a collection, it's an object. 
https://era.library.ualberta.ca/public/view/item/uuid:11274e20-0426-4e80-84f4-bef79dbd6633/

We can temporarily disable the community display to allow these items to show properly. And can probably generate a report on all the problematic issues for further investigation.

**Comment by sfarnel** [comment]
uuid:ca304e6a-06ab-419d-aa52-1771d3b641ee

is from collection

Current Index to Legal Periodicals
(uuid:828f5a2e-3f54-4399-8014-40d612d4920d)

and community

University of Alberta Libraries Licensed Resources
(uuid:11274e20-0426-4e80-84f4-bef79dbd6633)

these are active collections that should have been migrated.

interestingly, @weiweishi is right in that
uuid:11274e20-0426-4e80-84f4-bef79dbd6633
is also an ERA item, a stub record from 2010. It is an active item and so
may have been migrated also. @criedlberger, @piyapongch is the uuid meant
to be unique to object type or across the system?

this will cause an issue with other collections (see below) within that
community which are all ccid-protected:

uuid:25f42bd0-da6f-43bf-a14b-0088ff6d0eaa

uuid:172ac12c-c1d2-46c5-8389-18d303f31c9e

uuid:33e116e7-0a03-44e1-a886-a5f542db69ea

uuid:9630c6ac-bf4d-43c2-8b9c-587b6683864a

uuid:ad299eba-8f5e-4e07-a490-8a67adf9da27

uuid:caf9bb2f-1578-4074-bdeb-928fc2b7b40d

uuid:12c79e6e-28ea-4e1d-8312-d62115341501

uuid:859aa0b7-5f11-4775-8966-d218d3e007e6

uuid:260bce9a-4e84-421b-b5fc-5c791fa21975

uuid:cfba0bd8-d88f-4d2c-a559-46a478da117f

uuid:46662254-8570-49c3-9cf1-c1d61974a5e7

uuid:dbd9cdaa-693b-4674-bc5e-57431eb69bd5

uuid:0d9703de-172d-4264-98af-44acb1f3892e

uuid:53df6d40-5664-4a8c-8773-bd89b1a0b844

uuid:289284e9-354d-4b64-b8f4-57dfaff8e6ae

uuid:c594b61f-b417-421a-9046-1c6ab253b792

uuid:3f286112-3777-4158-b136-b439fdcf7514

uuid:761bdf34-7b34-4a86-a132-69d671d247f0

uuid:e56e93d1-b452-4735-ad2e-f56f6d8f251b

uuid:543305f4-cf10-4133-afc8-e1fe73ce91ad

uuid:9defcf0f-98bf-4bca-a0e8-312b0f4eb7b6

uuid:f5680006-5e71-40bf-bc27-d7897c29c1c4

uuid:4410fad4-79ec-4482-a6b5-75286e76669b

uuid:828f5a2e-3f54-4399-8014-40d612d4920d

uuid:5326c602-e22b-4efe-b7c5-fe77907290d3

uuid:c742bf70-09af-43da-bdef-b90549e2b015

uuid:e3b34864-cb7a-4ad8-b955-bb1c5dc8c5a7

uuid:9a140d28-8b5a-4e15-a2a8-af3a4b9ba535

uuid:e01c9b3a-06aa-454d-8614-228e99a36fb1

uuid:bfe70385-d3b6-4eac-8352-1bfdcd050def

uuid:ff9635a7-0951-4d7b-9195-c3b8e64c188d

uuid:5de12d25-1278-4f32-8539-ca84823396c2

uuid:6fd63225-1e30-4171-91fc-a07b5ef49191

uuid:71d0f5ad-54fe-4828-a969-11a0cabd7461

uuid:bd18d49e-ddab-4dad-b8a0-086d846057a8

uuid:50ab56cb-f549-43c8-8633-67cf81d2d91e

uuid:2f24cb02-65b2-4b75-8a3c-c51ec2085834

uuid:e2626e6b-590d-4010-ae44-021c4ed1a3f9

uuid:d811b1c5-6d07-4db7-9f3f-3956283c60bb

uuid:ed8825b1-6092-4c61-8ed2-63f2e1c741f1

uuid:ee6edfb4-4002-46ab-9c86-490d4f4b0af8

uuid:f3f969c2-c600-4c08-a2a2-0d9b0471ac35

uuid:5f9d6ffe-d547-4f0f-a6f4-6faf967b656e

uuid:507cc66c-37b0-42f7-a3ac-9f2610a83c30

uuid:928a345c-6c3f-4cdf-8543-d2362fcf9f0b

uuid:f4e8de21-bfd6-461a-94de-51d441fb0156

uuid:25287de9-fb3c-47fe-b12a-b46b494bbd1e

uuid:f1ebf8ef-4472-46e9-b1c0-e055f8678a5c

uuid:e301e3de-218a-41b9-bbd6-105e7fd63414

uuid:d2c9c9c1-6765-4f10-a809-ed65eb4d37c5

uuid:cff985d3-6340-4893-9c9e-da71bdee47e3

uuid:c97e1904-59e5-4766-942b-3db3acf0ce08

uuid:31912502-9cd1-47c1-9bfa-c58f157c95a2

uuid:8f80c8c0-3dda-4ada-a9cd-679c296bb173

uuid:0383c53f-7ef5-4bab-adf7-64e39f13e407

uuid:a7291c0d-cf8d-4861-bb1d-e9725f0840a5

uuid:8edc652d-e026-4e84-8045-831eba55b507

--

here is an item in a different community that is ccid-protected:

https://hydranorth.library.ualberta.ca/files/pc289k43m#.VhqU5aLwMiE

(uuid:db3eb892-249f-4239-bfb4-1ee28fdeb29d)

when logged in as admin I can search for it and see it. it is correctly
labelled as UofA Access. I can successfully download the item.

when not logged in if I paste
https://hydranorth.library.ualberta.ca/files/pc289k43m#.VhqU5aLwMiE into a
browser when not logged in I am prompted to log in and when I do I see the
item and can download it!

the remaining issue is that I'm not able to see the item metadata/search
for the item/browse for the item when not logged in




On Sun, Oct 11, 2015 at 1:38 AM, Weiwei Shi <notifications@github.com>
wrote:

> With the issue identified by @sfarnel <https://github.com/sfarnel> on the
> comment of #710 <https://github.com/ualbertalib/HydraNorth/issues/710>:
> e.g.: https://hydranorth.library.ualberta.ca/files/dn39x224x
>
> visibility shows as UofA. when I click on an item I get an error:
>
> "We're sorry, but something went wrong. If you are the application owner
> check the logs for more information."
>
> Log info:
> ActionView::Template::Error (Model mismatch. Expected Collection. Got:
> GenericFile):
> 8: <% @generic_file.belongsToCommunity.each do |c| %>
> 9:
> 10: <%= content_tag :td, class: 'attribute title', data: { id: c } do %>
> 11: <% community = Collection.find(c) %>
> 12: <%= link_to community.title, collections.collection_path(community) %>
> 13: <% end %>
> 14:
>
> Some issue with a particular item:
> uuid:ca304e6a-06ab-419d-aa52-1771d3b641ee
>
> The community in the foxml has an
> uuid:11274e20-0426-4e80-84f4-bef79dbd6633
> However, this is not a collection, it's an object.
>
> https://era.library.ualberta.ca/public/view/item/uuid:11274e20-0426-4e80-84f4-bef79dbd6633/
>
> We can temporarily disable the community display to allow these items to
> show properly. And can probably generate a report on all the problematic
> issues for further investigation.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ualbertalib/HydraNorth/issues/712>.
>



-- 
Sharon Farnel
Metadata Coordinator
University of Alberta Libraries
sharon.farnel@ualberta.ca
780-492-3685

**Comment by weiweishi** [comment]
The item is also in the migration batch, and was migrated as a object. I've fixed this community and will fix the community information on children objects. We need to better understand the issue of duplicated uuid, as there may be other similar situation elsewhere.

**Comment by weiweishi** [comment]
Realized the issue was with the last-minute deletion happened before the Oct 6 launch. The collection was deleted then, and the object was then migrated after. I'm now fixing all the objects that are associated with the deleted collection.

**Comment by sfarnel** [comment]
Found another example of duplicate uuids:

https://era.library.ualberta.ca/public/view/item/uuid:20b206ff-36d4-46f9-b3f3-3376e496f863/

and

https://era.library.ualberta.ca/public/view/collection/uuid:20b206ff-36d4-46f9-b3f3-3376e496f863

We'll try and generate a list of these to pass along for fixing.

Cheers,
Sharon

On Tue, Oct 13, 2015 at 10:46 AM, Weiwei Shi <notifications@github.com>
wrote:

> Realized the issue was with the last-minute deletion happened before the
> Oct 6 launch. The collection was deleted then, and the object was then
> migrated after. I'm now fixing all the objects that are associated with the
> deleted collection.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ualbertalib/HydraNorth/issues/712#issuecomment-147773164>
> .
>



-- 
Sharon Farnel
Metadata Coordinator
University of Alberta Libraries
sharon.farnel@ualberta.ca
780-492-3685

**Comment by weiweishi** [comment]
This one probably is okay, as the collection was not deleted during our
rush deletion on Oct 6. When search the uuid in Hydranorth, only the
collection is migrated.

*Weiwei Shi*

Digital Initiative Applications Librarian
University of Alberta Libraries
2-10L Cameron Library
Edmonton, Alberta, Canada T6G 2J8
Phone:(780)492-7802
Fax: (780)248-1209
Email: weiwei.shi@ualberta.ca

**Comment by anayram** [comment]
In relation to duplicate uuids: it may be that it is not a case of duplication but of different views available in ERA for a specific foxml item.

For example, the community below can be viewed in old ERA both as community and as item but it is the same foxml:
https://era.library.ualberta.ca/public/view/community/uuid:11274e20-0426-4e80-84f4-bef79dbd6633/
and
https://era.library.ualberta.ca/public/view/item/uuid:11274e20-0426-4e80-84f4-bef79dbd6633/

This applies to all communities and collections (i.e. any community can be seen as item and any collection can be seen as item, but no community can be seen as collection or viceversa).

**Comment by pbinkley** [comment]
That would lead to hydranorth being able to fetch foxml from old ERA both
for a collection and for an item with the same uuid, without there being
actual duplication in old ERA.

**Comment by anayram** [comment]
Does the migration script filter duplicate items coming in by uuid regardless of whether it is a generic item or a community/collection?

Asking because from the migration package we excluded:
deleted items
inactive legacy theses
system files
but I did not filter out collections or communities (thinking that, because they were migrated first, they would have been caught as duplicates).

Just wondering if this has anything to do with the fact that a community or collection was migrated as generic file, summed with the fact that hydranorth seems to be able to fetch foxml from old ERA both for a collection and an item with the same uuid, like @pbinkley says.

**Comment by weiweishi** [comment]
It will check for uuid duplicates regardless of their models, so generally
speaking, including the collections/communities in the migration package
should not cause any problem. Collections/Communities would be migrated
first, and any items with the same uuid would treated as a duplication, and
would not be migrated.

This particular issue happened because we did the last-minute deletion on
Oct 6 on inactive/embargo items, and this collection was deleted then. And
then the object with the same uuid was migrated, because the collection has
been previously removed.



*Weiwei Shi*

Digital Initiative Applications Librarian
University of Alberta Libraries
2-10L Cameron Library
Edmonton, Alberta, Canada T6G 2J8
Phone:(780)492-7802
Fax: (780)248-1209
Email: weiwei.shi@ualberta.ca

**Comment by anayram** [comment]
Awesome, thank you Weiwei.

**Comment by weiweishi** [comment]
This particular issue with community "University of Alberta Licensed Resources" should be fixed.

**Comment by pbinkley** [comment]
I've checked the problem items that were reported above and spot-checked the list of potential problem items that @sfarnel provided above, and I haven't found any problems. I think the problem has been solved by @weiweishi 's remigrating of the deleted community and deletion of the item that was migrated from a community. 
The only remaining problem is that some items (e.g. uuid:987628e1-5cb5-473e-9d86-e003d7f7f36b/) belonging to some collections (e.g. uuid:25287de9-fb3c-47fe-b12a-b46b494bbd1e) do not appear when you visit the collection page (https://hydranorth.library.ualberta.ca/collections/44558s12j), but this is presumably due to #729 and not to any problem in this issue. I think we have to keep this issue open until #729 closes and we can confirm that the items appear under their collection.
